### ISSUE TITLE
getcwd syscall does not work in rclone mounted drives

### ISSUE THREAD

**Issue by Jamesernator** [rclone/rclone#4104]
#### What is the problem you are having with rclone?

Programs that use the `getcwd()` syscall fail when inside an rclone mount with `ENOENT`.

For example I go into a rclone mounted drive `cd ~/dropbox/Documents`, then run something like `node -p "process.cwd()"`, this fails with `ENOENT`.

I even made a basic C script with nothing but the syscall:

```c
#include <unistd.h>
#include <stdio.h>
#include <limits.h>

int main() {
  char cwd[PATH_MAX];
  if (getcwd(cwd, sizeof(cwd)) != NULL) {
    printf("%s\n", cwd);
  } else {
    perror("getcwd() error");
    return 1;
  }
  return 0;
}
```

Which also fails, so this confirms it's the syscall failing and not something in Node/Python.

NOTE: This only fails in a subdirectory, it does not fail in the mount directory itself e.g. `~/mount/` is fine, but `~/mount/subdirectory/` is not.

#### What is your rclone version (output from `rclone version`)

1.51.0

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Ubuntu 18.04 and 19.10 both fail.

####  Which cloud storage system are you using? (eg Google Drive)

Multiple, failed for WebDAV (nextcloud), Dropbox and Google Drive.

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

The WebDAV (nextcloud) is mounted as:

```bash
/usr/bin/rclone mount nextcloud:/ /home/jamesernator/nextcloud
```

The google drive is mounted as:

```bash
/usr/bin/rclone mount googleDrive:/ /home/jamesernator/googleDrive --dir-cache-time 30s --cache-dir /home/jamesernator/.rclone-mount-cache/ --vfs-cache-max-age 720h0m0s --vfs-cache-mode full --vfs-cache-max-size 2G
```

And dropbox is mounted using a [cache remote](https://rclone.org/cache/) with:

```bash
/usr/bin/rclone mount dropbox-cache:/ /home/jamesernator/dropbox
```

NOTE: All are mounted automatically through systemd services, however manually mounting does not fix the situation.

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

This logs contains the results of running `rclone mount -vv "nextcloud:/" ~/nextcloud-test/ 2> out.log" followed by the following two commands in `zsh` (in a separate shell):

```zsh
cd nextcloud-test/Documents
node -e 'process.cwd()'
```

[out.log](https://github.com/rclone/rclone/files/4419026/out.log)

**Comment by Jamesernator** [comment]
Doing a tiny bit more digging, it seems that it works okay when running in a `bash` shell for some reason?

e.g. `node -p 'process.cwd()'` succeeds when run in bash but not in zsh.

**Comment by ncw** [comment]
I've been trying to replicate this using ubuntu 19.10 with zsh and it is stubbornly refusing to go wrong!

Which kernel are you using? I'm using `5.3.0-42-generic` with zsh 5.7.1-1ubuntu2.

Can you strace your C program going wrong for me?

Is there anything else about your system which is unusual?

**Comment by Jamesernator** [comment]
Kernel on main machine is: `5.3.0-45-generic`
Also affects my other machine which is: `5.3.0-42-generic`

Only thing I can thing of in common that might be significant is they both have LUKS encrypted file systems.

Here's an `strace` of `cwd`:

[out.log](https://github.com/rclone/rclone/files/4423923/out.log)

**Comment by ncw** [comment]
> Kernel on main machine is: `5.3.0-45-generic`
> Also affects my other machine which is: `5.3.0-42-generic`

My machine is `5.3.0-42-generic` and is also running on LUKS encrypted file system.

What about zsh version?

Do you have lots of stuff in your zsh rc?

I'm starting a zsh from bash by typing zsh - is that OK?

Maybe you could try reproducing this in an ubuntu VM image I can download too?

**Comment by Jamesernator** [comment]
> What about zsh version?

zsh 5.7.1 (x86_64-ubuntu-linux-gnu)

> Do you have lots of stuff in your zsh rc?

Yes, but it also happens when running `zsh -f` which doesn't load them.

> I'm starting a zsh from bash by typing zsh - is that OK?

I'm not sure, but doing a bit of investigation, the directory in when switching from `zsh` to `bash` matters:

(Both of these examples behave the same when using shell replacement via `exec bash`, rather than subshell).

This fails:

```
zsh> cd ~/nextcloud/Documents/
zsh> bash
bash> cwd
getcwd() error: No such file or directory
```

But this doesn't:

```
zsh> bash
bash> cd ~/nextcloud/Documents/
bash> cwd
/home/jamesernator/nextcloud/Documents/
```

> Maybe you could try reproducing this in an ubuntu VM image I can download too?

Tried creating a VM, but the issue didn't seem to happen. If you want I can create a user for you on my machine if you want.

**Comment by ncw** [comment]
OK I managed to replicate that :-)

It seems that starting the shell in the subdirectory of an `rclone mount` is the critical part. It goes wrong with`bash` and `zsh` for me like that.

```
$ cd /mnt/tmp/sub/
bash
$ ~/Rclone/pwd
getcwd() error: No such file or directory
$ /bin/pwd
/
```

pwd is the little C program you made. However this does work with `/bin/pwd`

Looking at the strace of `/bin/pwd` you can see it gets the same ENOENT from `getcwd`

However it then appears to do a workaround which involves listing `..` and finding the current directory in it.

```
getcwd(0x5612436a6500, 4096)            = -1 ENOENT (No such file or directory)
lstat("/", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
stat(".", {st_mode=S_IFDIR|0775, st_size=0, ...}) = 0
openat(AT_FDCWD, "..", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 3
fstat(3, {st_mode=S_IFDIR|0775, st_size=0, ...}) = 0
fchdir(3)                               = 0
fstat(3, {st_mode=S_IFDIR|0775, st_size=0, ...}) = 0
getdents64(3, /* 10 entries */, 32768)  = 304
close(3)                                = 0
openat(AT_FDCWD, "..", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 3
fstat(3, {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
fchdir(3)                               = 0
fstat(3, {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
getdents64(3, /* 8 entries */, 32768)   = 216
lstat("data", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
[snip]
lstat("tmp", {st_mode=S_IFDIR|0775, st_size=0, ...}) = 0
close(3)                                = 0
openat(AT_FDCWD, "..", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 3
fstat(3, {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
fchdir(3)                               = 0
fstat(3, {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
getdents64(3, /* 35 entries */, 32768)  = 944
close(3)                                = 0
fstat(1, {st_mode=S_IFCHR|0620, st_rdev=makedev(0x88, 0x23), ...}) = 0
write(1, "/mnt/tmp/sub\n", 13)          = 13
```

I tried this with `rclone cmount` which is a version of rclone mount which uses libfuse and this did **not** go wrong!

There is a clue here... from within the subdirectory of the freshly started bash I get this - no `.` or `..` entries.

```
$ ls -la
total 0
```

In fact they seem to be missing for all the directories :confused: 

Checking old versions of rclone I see they have never had `.` or `..`

I tried adding `.` and `..` entries and it didn't make any difference!

I think the problem comes from here

```
$ ls -l /proc/self/cwd
lrwxrwxrwx 1 ncw ncw 0 Apr 10 16:28 /proc/self/cwd -> '/mnt/tmp/sub (deleted)'
```

Why does the kernel think `/mnt/tmp/sub` is deleted?

I think this little sequence sums it up

```
$ cd /mnt/tmp/sub/
$ ls -l /proc/self/cwd
lrwxrwxrwx 1 ncw ncw 0 Apr 10 16:33 /proc/self/cwd -> /mnt/tmp/sub
$ bash
$ ls -l /proc/self/cwd
lrwxrwxrwx 1 ncw ncw 0 Apr 10 16:33 /proc/self/cwd -> '/mnt/tmp/sub (deleted)'
```

I tried increasing the `--attr-timeout 10s` and then the `ls -l /proc/self/cwd` worked for 10s then started failing as did the pwd.c code. So the kernel appears to think the directory had been deleted when it drops out of the attribute cache.

Maybe I'm supplying a wrong parameter or maybe it is a bug in the kernel...

**Comment by dragetd** [comment]
I am experiencing the same issues right now. On Ubuntu 18.04

Kernel: 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

I was confused why my dolphin randomly assumed directories to be deleted and some python scripts failing in subdirectories with a cwd call. Looks pretty much like the issue in question, but on an older kernel.

I'll try older rclone versions and see if this is a recent regression.

**Comment by ncw** [comment]
Hmm, I wonder if this was caused by a kernel change. 4.15 is quite old though.

Does adding `--attr-timeout 10s` make it work for 10 seconds?

**Comment by dragetd** [comment]
Oh, neat. Fortune favors fools, I guess!

I thought I could help debugging this but just found no clue whatsoever, testing different versions. The --attr-timeout parameter did also affect my case, but I wanted to make sure and tested again today. Only to find out it magically works!

Why? Since I also checkout out the repository yesterday which happens cause (master branch) compiled version in my GOPATH, I was running master.

So, the problem seems to be fixed in master. Out of curiosity I want to bisect the commit causing it, something I am better at than go. :D

**Comment by ncw** [comment]
Hmm, I think I might need to ask a proper file system developer for help here! I think the solution is something to do with the attribute timeouts on the inodes and invalidating them properly but I'm a bit out of my depth here!

**Comment by dciangot** [comment]
Hi there, is there any news on this? We have the same issue when trying to use `rclone mount` to use a remote s3/webdav endpoint as a home directory for jupyter notebook.

**Comment by ncw** [comment]
I haven't figured out what is going on yet...

I submitted https://github.com/bazil/fuse/issues/250 in the hope that someone more knowledgeable about FUSE internals might help :-)

**Comment by ncw** [comment]
A massive thank you to @tv42 who has worked out what is going on! See bazil/fuse#250 for the full explanation.

It turned out to be a problem in rclone. It was sending different nodes to the fuse layer with the same information in which caused fuse to believe that the directory had been deleted and re-created. This was fixed by caching the actual node for an item and returning that.

Please have a go with this and tell me what you think.

https://beta.rclone.org/branch/v1.51.0-264-g2696e369-fix-4104-cwd-disappears-beta/ (uploaded in 15-30 mins)

**Comment by dciangot** [comment]
I can confirm that at least for our use case everything is working with that fix. 
E.g. python os.getcwd and also git clone that wasn't working before from within a mounted folder, now are ok.
Thanks a lot!

**Comment by darthShadow** [comment]
@ncw Could this also be the cause of the memory leaks related to `--attr-timeout`, when set to 0, or is that still a different issue?

**Comment by ncw** [comment]
@dciangot thanks for testing

@darthShadow yes that crossed my mind too... Fancy giving it a test?

**Comment by darthShadow** [comment]
Will do. I assume this is still the best way to reproduce it: https://github.com/rclone/rclone/issues/2157#issuecomment-374755748?

**Comment by ncw** [comment]
That looks like a good thing to try

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.52

> Will do. I assume this is still the best way to reproduce it: [#2157 (comment)](https://github.com/rclone/rclone/issues/2157#issuecomment-374755748)?

You need to add `--attr-timeout 0` to the mount command. I tried it and it does indeed appear to be fixed :-)

**Comment by darthShadow** [comment]
> I tried it and it does indeed appear to be fixed

Superfast as always... :stuck_out_tongue:

Hadn't gotten a chance to try it yet. Just set it on both my active mounts.

**Comment by dragetd** [comment]
Since I forgot to post it here: It also fixed it for me.

If @Jamesernator could confirm, this would be nice. Otherwise this one can be closed I believe. Awesome thing, thanks for the fix! :)

**Comment by ncw** [comment]
The fix for this went into 1.52 so it should be all good now :-)
